File: minigit-5/ts/src/commands/cat-file.ts
// cat-file command - Examine objects

import { findRepoRoot } from '../utils';
import { resolveRevision } from '../refs';
import { readObject, getCommit, getTree, parseTag, expandShortSha } from '../objects';

export function catFile(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let showType = false;
  let showSize = false;
  let prettyPrint = false;
  let objectType: string | null = null;
  let objectRef: string | null = null;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-t') {
      showType = true;
    } else if (arg === '-s') {
      showSize = true;
    } else if (arg === '-p') {
      prettyPrint = true;
    } else if (['blob', 'tree', 'commit', 'tag'].includes(arg)) {
      objectType = arg;
    } else if (!arg.startsWith('-')) {
      objectRef = arg;
    }
  }

  if (!objectRef) {
    console.error('fatal: object reference required');
    return 1;
  }

  // Resolve object
  let sha: string;
  try {
    sha = resolveRevision(objectRef, repoRoot);
  } catch {
    // Try as raw SHA
    try {
      sha = expandShortSha(objectRef, repoRoot);
    } catch (e: any) {
      console.error(`fatal: Not a valid object name ${objectRef}`);
      return 1;
    }
  }

  // Read object
  let obj: { type: string; content: Buffer };
  try {
    obj = readObject(sha, repoRoot);
  } catch (e: any) {
    console.error(`fatal: Not a valid object name ${objectRef}`);
    return 1;
  }

  const { type, content } = obj;

  // Check type if specified
  if (objectType && type !== objectType) {
    console.error(`fatal: expected ${objectType}, got ${type}`);
    return 1;
  }

  // Output based on flags
  if (showType) {
    console.log(type);
    return 0;
  }

  if (showSize) {
    console.log(content.length.toString());
    return 0;
  }

  if (prettyPrint || objectType) {
    switch (type) {
      case 'blob':
        process.stdout.write(content);
        break;

      case 'tree':
        const entries = getTree(sha, repoRoot);
        for (const entry of entries) {
          const entryType = entry.mode === '40000' ? 'tree' : 'blob';
          // Pad mode to 6 chars for regular files
          const modeStr = entry.mode === '40000' ? '040000' : entry.mode;
          console.log(`${modeStr} ${entryType} ${entry.sha}\t${entry.name}`);
        }
        break;

      case 'commit':
        const commit = getCommit(sha, repoRoot);
        console.log(`tree ${commit.tree}`);
        for (const parent of commit.parents) {
          console.log(`parent ${parent}`);
        }
        console.log(`author ${commit.author}`);
        console.log(`committer ${commit.committer}`);
        console.log('');
        console.log(commit.message);
        break;

      case 'tag':
        const tag = parseTag(content);
        console.log(`object ${tag.object}`);
        console.log(`type ${tag.objectType}`);
        console.log(`tag ${tag.tagName}`);
        console.log(`tagger ${tag.tagger}`);
        console.log('');
        console.log(tag.message);
        break;

      default:
        console.error(`Unknown object type: ${type}`);
        return 1;
    }
    return 0;
  }

  // Raw output
  process.stdout.write(content);
  return 0;
}

File: minigit-5/py/commands/cat_file.py
"""cat-file command - Examine object internals."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, read_object
from objects import Blob, Tree, Commit, Tag
from refs import resolve_revision


def run(args: list[str]) -> int:
    """Examine object internals."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    if len(args) < 2:
        print("usage: minigit cat-file (-t|-s|-p) <object>", file=sys.stderr)
        return 1

    # Parse arguments
    mode = args[0]
    obj_ref = args[1]

    # Handle positional usage: cat-file blob <sha>
    if mode in ('blob', 'tree', 'commit', 'tag'):
        expected_type = mode
        obj_ref = args[1]
        return show_raw(repo_root, obj_ref, expected_type)

    # Resolve object
    sha = resolve_revision(repo_root, obj_ref)
    if sha is None:
        print(f"fatal: not a valid object name: {obj_ref}", file=sys.stderr)
        return 1

    try:
        obj_type, data = read_object(sha, repo_root)
    except Exception as e:
        print(f"fatal: not a valid object: {sha}", file=sys.stderr)
        return 1

    if mode == '-t':
        print(obj_type)
    elif mode == '-s':
        print(len(data))
    elif mode == '-p':
        return pretty_print(repo_root, sha, obj_type, data)
    else:
        print(f"unknown option: {mode}", file=sys.stderr)
        return 1

    return 0


def show_raw(repo_root: Path, obj_ref: str, expected_type: str) -> int:
    """Show raw object content."""
    sha = resolve_revision(repo_root, obj_ref)
    if sha is None:
        print(f"fatal: not a valid object name: {obj_ref}", file=sys.stderr)
        return 1

    obj_type, data = read_object(sha, repo_root)
    if obj_type != expected_type:
        print(f"fatal: {sha} is not a {expected_type}", file=sys.stderr)
        return 1

    if expected_type == 'blob':
        try:
            print(data.decode('utf-8'), end='')
        except:
            sys.stdout.buffer.write(data)
    else:
        print(data.decode('utf-8'), end='')

    return 0


def pretty_print(repo_root: Path, sha: str, obj_type: str, data: bytes) -> int:
    """Pretty-print object content."""
    if obj_type == 'blob':
        try:
            print(data.decode('utf-8'), end='')
        except:
            sys.stdout.buffer.write(data)

    elif obj_type == 'tree':
        tree = Tree.read(sha, repo_root)
        for entry in tree.entries:
            type_name = 'tree' if entry.mode == 0o40000 else 'blob'
            print(f"{entry.mode:06o} {type_name} {entry.sha}\t{entry.name}")

    elif obj_type == 'commit':
        commit = Commit.read(sha, repo_root)
        print(f"tree {commit.tree_sha}")
        for parent in commit.parents:
            print(f"parent {parent}")
        print(f"author {commit.author} <{commit.author_email}> {commit.author_date}")
        print(f"committer {commit.committer} <{commit.committer_email}> {commit.committer_date}")
        print()
        print(commit.message)

    elif obj_type == 'tag':
        tag = Tag.read(sha, repo_root)
        print(f"object {tag.object_sha}")
        print(f"type {tag.object_type}")
        print(f"tag {tag.tag_name}")
        print(f"tagger {tag.tagger} <{tag.tagger_email}> {tag.tagger_date}")
        print()
        print(tag.message)

    return 0
